2/contrib/jenkins/install-plugins.sh

```diff
@@ -233,23 +235,45 @@ function resolveDependencies() {
     for d in "${array[@]}"
     do
         plugin="$(cut -d':' -f1 - <<< "$d")"
+	#
+	# Note, matrix-auth plugin notes cloudbees-folder as optional in the archive, but then failed to
+	# load, citing a dependency that is too old, during testing ... so we will download optional dependencies
+	#
+	local versionFromPluginParam
         if [[ $d == *"resolution:=optional"* ]]; then
-            echo "Skipping optional dependency $plugin"
-        else
-            local pluginInstalled
-            if pluginInstalled="$(echo "${bundledPlugins}" | grep "^${plugin}:")"; then
-                pluginInstalled="${pluginInstalled//[$'\r']}"
-                local versionInstalled; versionInstalled=$(versionFromPlugin "${pluginInstalled}")
-                local minVersion; minVersion=$(versionFromPlugin "${d}")
-                if versionLT "${versionInstalled}" "${minVersion}"; then
-                    echo "Upgrading bundled dependency $d ($minVersion > $versionInstalled)"
-                    download "$plugin" &
-                else
-                    echo "Skipping already bundled dependency $d ($minVersion <= $versionInstalled)"
-                fi
+            #echo "Skipping optional dependency $plugin"
+	    versionFromPluginParam="$(cut -d';' -f1 - <<< "$d")"
+	else
+	    versionFromPluginParam=$d
+	fi
+        local pluginInstalled
+        local minVersion; minVersion=$(versionFromPlugin "${versionFromPluginParam}")
+
+	set +o pipefail
+	local filename; filename=$(getArchiveFilename "$plugin")
+	local previouslyDownloadedVersion; previouslyDownloadedVersion=$(get_plugin_version $filename)
+	set -o pipefail
+	    
+        if pluginInstalled="$(echo "${bundledPlugins}" | grep "^${plugin}:")"; then
+            pluginInstalled="${pluginInstalled//[$'\r']}"
+            local versionInstalled; versionInstalled=$(versionFromPlugin "${pluginInstalled}")
+            if versionLT "${versionInstalled}" "${minVersion}"; then
+                echo "Upgrading bundled dependency $d ($minVersion > $versionInstalled)"
+                download "$plugin" "$minVersion" "true"
             else
-                download "$plugin" &
+                echo "Skipping already bundled dependency $d ($minVersion <= $versionInstalled)"
             fi
+	    continue
+        fi
+
+	if [[ ! $previouslyDownloadedVersion ]]; then
+	    echo "Downloading previously uninstalled dependency plugin $plugin version $minVersion"
+            download "$plugin" "$minVersion"
+	else
+	    if versionLT "${previouslyDownloadedVersion}" "${minVersion}"; then
+		echo "Upgrading previously downloaded plugin $plugin at $previouslyDownloadedVersion to $minVersion"
+		download "$plugin" "$minVersion" "true"
+	    fi
         fi
```

Comment:
i'm confused by this whole if/else block.  it appears to be downloading the plugin for which we are currently resolving dependencies.  but should we be doing that here?  The top level plugin itself is downloaded elsewhere.  Isn't it sufficient to just deal w/ downloading the minversion of the dependencies in this function?